src/Winton.DomainModelling.DocumentDb/ValueObjectFacade.cs

```diff
@@ -0,0 +1,120 @@
+    public sealed class ValueObjectFacade : IValueObjectFacade
+    {
+        private readonly Database _database;
+        private readonly IDocumentClient _documentClient;
+        private readonly DocumentCollection _documentCollection;
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="ValueObjectFacade" /> class.
+        /// </summary>
+        /// <param name="database">The DocumentDb database.</param>
+        /// <param name="documentCollection">The DocumentDb collection. Partitioned collections are not supported.</param>
+        /// <param name="documentClient">A document client implementation.</param>
+        public ValueObjectFacade(
+            Database database,
+            DocumentCollection documentCollection,
+            IDocumentClient documentClient)
+        {
+            if (documentCollection.PartitionKey.Paths.Any())
+            {
+                throw new NotSupportedException("Partitioned collections not supported.");
+            }
+
+            _database = database;
+            _documentCollection = documentCollection;
+            _documentClient = documentClient;
+        }
+
+        /// <inheritdoc />
+        /// <summary>
+        ///     Create a Value Object of a specified type.
+        /// </summary>
+        /// <typeparam name="TValueObject">The type of the Value Object.</typeparam>
+        /// <param name="valueObject">The Value Object to persist.</param>
+        /// <returns>A Task.</returns>
+        public async Task Create<TValueObject>(TValueObject valueObject)
+            where TValueObject : struct, IEquatable<TValueObject>
+        {
+            string valueObjectType = ValueObjectDocument<TValueObject>.GetDocumentType();
+
+            ValueObjectDocument<TValueObject> document =
+                _documentClient.CreateDocumentQuery<ValueObjectDocument<TValueObject>>(GetUri())
+                               .Where(x => x.Type == valueObjectType)
+                               .AsEnumerable()
+                               .SingleOrDefault(x => x.ValueObject.Equals(valueObject));
+
+            if (document == null)
+            {
+                document = new ValueObjectDocument<TValueObject>(valueObject);
+
+                await _documentClient.CreateDocumentAsync(GetUri(), document);
+            }
+        }
+
+        /// <inheritdoc />
+        /// <summary>
+        ///     Delete a Value Object of a specified type.
+        /// </summary>
+        /// <typeparam name="TValueObject">The type of the Value Object.</typeparam>
+        /// <param name="valueObject">The Value Object to delete.</param>
+        /// <returns>A Task.</returns>
+        public async Task Delete<TValueObject>(TValueObject valueObject)
+            where TValueObject : struct, IEquatable<TValueObject>
+        {
+            string valueObjectType = ValueObjectDocument<TValueObject>.GetDocumentType();
+
+            ValueObjectDocument<TValueObject> document =
```

Comment:
Given this is the same as for the `Create` method, I think we could have a private `Find` method that finds the `ValueObjectDocument` by the `TValueObject`.

Reply:
I think you could also extract:

```
string valueObjectType = ValueObjectDocument<TValueObject>.GetDocumentType();
ValueObjectDocument<TValueObject> document =
    _documentClient.CreateDocumentQuery<ValueObjectDocument<TValueObject>>(GetUri())
                   .Where(x => x.Type == valueObjectType)
                   .AsEnumerable()
```

into a private method called something like `CreateValueObjectDocumentQuery`